### ISSUE TITLE
add in-scheme concept relations

### ISSUE THREAD

**Issue by dwinston** [polyneme/heliokos#19]
Add in-scheme concept relationships, that is, the ability to assert a qualified directed edge between two concepts in the context of a Concept Scheme. The two relationships I'd like to enable are "is a broader concept than" (represented as the string `narrower`), and "is related to" (represented as the string `related`), with the option to add more (not anytime soon, though).

So, for example, on a concept scheme page with two concepts A and B, there are four possible relationships I can assert: A-narrower-B, B-narrower-A, A-related-B, B-related-A. UX-wise, I think I'd like three select elements, one for the subject concept, one for the predicate, and one for the object concept. Can be a simple select for now, with an eye towards it allowing typeahead/auto-complete if there are many concepts and relationships.

There will be backend logic to supply what the front-end needs to accept or reject relationship assertions. For example, if a user has added A-narrower-B, they cannot assert B-narrower-A (unless they remove their first assertion).

**Comment by dwinston** [comment]
branch `in-scheme-concept-relations` pushed for #20 draft PR

**Comment by cferdinandi** [comment]
@dwinston Awesome, thanks for the details!

Just to make sure I fully understand, we'll add a second form to the `/conceptschemes/{id}` route (which currently has the "add a concept" form).

It will have three fields...

1. "First concept" (select menu)
2. "Relationship..." (select menu with options like "narrower", "related", etc.)
3. "Second concept" (select menu)

Is that right?

**Comment by dwinston** [comment]
@dwinston exactly.

**Comment by dwinston** [comment]
@cferdinandi exactly.

**Comment by cferdinandi** [comment]
@dwinston Looking at the PR you have setup, are these the correct field names to use?

- `subject_concept_id`: The first concept
- `object_concept_id`: The second concept
- `relation`: How the first and second concept are related?

And the form will `POST` to `/conceptscheme/{scheme id}/relation`?

**Comment by dwinston** [comment]
Yes, that looks right

**Comment by cferdinandi** [comment]
Branch `scheme-concept-relations-fed` pushed for PR https://github.com/polyneme/heliokos/pull/21 that establishes a front end UI for this

**Comment by dwinston** [comment]
going to close in favor of more minimal, precise issues related to adding in-scheme concept relations

**Comment by cferdinandi** [comment]
@dwinston Going in a different direction with this?

**Comment by dwinston** [comment]
no. and re-thinking it, I can wrap my head around this, so even though it may take more than one PR to close, let's keep it open. 😄

**Comment by dwinston** [comment]
@cferdinandi I added Python code to return

(1) a list of a concept scheme's relations (`scheme.relations`), where each list element is [subject_concept_id, relation, object_concept_id], and 

(2) a list of relations that a concept scheme will deny (scheme.deny_relations), where each element is as above.

I have not made endpoints for these yet, nor refactored the concept scheme template to include them in the html, but hopefully these give you an idea of how to close this issue.

I'll want a scheme page to list out its current relations, and to use `deny_relations` to rule out subjects, relations, and objects (and relations and objects again once a subject is selected, etc.).

**Comment by cferdinandi** [comment]
@dwinston Great, thanks!

On an individual scheme page, will be listing the relations somewhere, or is that it's own dedicated page? If you instead need/want some UI/UX guidance around that, let me know.

I think this is where you're headed with this already, but in an ideal world, when the user submits the form, the server returns a full HTML file, with whatever content you want rendered/updated into the current (if any) on success, and an error message of some sort on failure.

I can update the `<hx-form>` element to support grabbing error messages from the HTML on failure.

This approach ensures that even when JS fails to load, the app is fully usable.

**Comment by dwinston** [comment]
> On an individual scheme page, will be listing the relations somewhere, or is that it's own dedicated page? If you instead need/want some UI/UX guidance around that, let me know.

I was thinking a listing on the scheme page at first, to eventually be replaced with / supplemented by markup for a [mermaid flowchart with clickable nodes](https://mermaid.js.org/syntax/flowchart.html#interaction) so that one can visualize the scheme, or at least part of it (perhaps just the top concepts and their immediate relations), with clicking on a node prompting a refresh of the diagram centered on that node and its backward and forward relations (for which it is the subject or object).

**Comment by dwinston** [comment]
> I think this is where you're headed with this already, but in an ideal world, when the user submits the form, the server returns a full HTML file, with whatever content you want rendered/updated into the current (if any) on success, and an error message of some sort on failure.

Yes

**Comment by dwinston** [comment]
> I can update the <hx-form> element to support grabbing error messages from the HTML on failure.

great

**Comment by dwinston** [comment]
> This approach ensures that even when JS fails to load, the app is fully usable.

Yes. So do you want me to send an HTML snippet if your code sends a special HTTP header? Or do you think its fine to just return the whole page and the JS will pick out what it needs if it makes an async call?

**Comment by dwinston** [comment]
a site that demonstrates the use of mermaid flowchart diagrams to visualize graph fragments and navigate a graph interactively: <https://d3fend.mitre.org/dao/artifact/d3f:Software/> (I had no involvement; I just haven't been able to find a similar example on the web).

**Comment by cferdinandi** [comment]
> Yes. So do you want me to send an HTML snippet if your code sends a special HTTP header? Or do you think its fine to just return the whole page and the JS will pick out what it needs if it makes an async call?

Given how the `<hk-form>` code is setup, you can send an enter HTML page, IMO. Probably provides better progressive enhancement/non-JS experience, too.

**Comment by cferdinandi** [comment]
@dwinston Just checking where we're at with this? Any updates needed on my end at this time?

**Comment by dwinston** [comment]
@cferdinandi thanks for poking. I'll try to push code to enable concrete next steps, raise issues for them, and set priorities, by COB today.

**Comment by dwinston** [comment]
@cferdinandi okay, this is ready for you to implement UI/UX. I include relations in the scheme display, and I include a payload of `deny_relations` to disallow. I display the latter, but only to bring your attention to the relevant template variable and its form. The backend returns a 422 error when a disallowed relation is attempted via POST, but I'd like the front-end to prevent this through dynamic disabling of select options.

**Comment by cferdinandi** [comment]
@dwinston - The `{{ deny_relations }}` string is returning with single quotes around it. In order to do the JS manipulation I need to, it needs double quotes to count as valid JSON. Is there a way to make that happen?

**Comment by dwinston** [comment]
> an one concept be both broader and related to another concept in a scheme?

No. This should be disallowed.

**Comment by cferdinandi** [comment]
@dwinston Excellent, that simplifies things! So basically, if a relationship already exists between two values, we shouldn't allow another one to be established?

**Comment by dwinston** [comment]
right (at least for the `narrower` and `related` relations. which is what I plan to restrict to).

**Comment by dwinston** [comment]
https://www.w3.org/TR/skos-reference/#L4261

> not only are the immediate (i.e., direct) hierarchical and associative links disjoint, but associative links are also disjoint with indirect hierarchical links.

hierarchical: `narrower`
associative: `related`

so if A-- narrower -->B and B-- narrower -->C, we want to disallow A-- related -->C (which I hope I am doing in my deny_relations payload.
